specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/webapplicationfirewall.json

```diff
@@ -667,25 +675,124 @@
-            "Contains",
-            "StartsWith",
-            "EndsWith",
-            "EqualsAny"
-          ],
           "description": "When matchVariable is a collection, operate on the selector to specify which elements in the collection this exclusion applies to.",
-          "x-ms-enum": {
-            "name": "OwaspCrsExclusionEntrySelectorMatchOperator",
-            "modelAsString": true
-          }
+          "$ref": "#/definitions/MatchOperator"
         },
         "selector": {
           "type": "string",
           "description": "When matchVariable is a collection, operator used to specify which elements in the collection this exclusion applies to."
         }
       }
+    },
+    "MatchOperator": {
+      "type": "string",
+      "enum": [
+        "Equals",
+        "Contains",
+        "StartsWith",
+        "EndsWith",
+        "EqualsAny"
+      ],
+      "description": "Defines the logic for comparing two strings.",
+      "x-ms-enum": {
+        "name": "OwaspCrsExclusionEntrySelectorMatchOperator",
+        "modelAsString": true
+      }
+    },
+    "BotDetection": {
+      "description": "Allows the detection of bots and specify action for each bot category.",
+      "required": [
+        "good", 
+        "neutral", 
+        "bad"
+      ],
+      "properties": {
+        "good": {
+          "description": "Describes the action to be taken when detecting a good bot.",
+          "properties": {
+            "action": {
+              "description": "Describes which action should be triggered by the bot category.",
+              "$ref": "#/definitions/Action"
+            }
+          }
+        },
+        "neutral": {
+          "description": "Describes the action to be taken when detecting a neutral bot.",
+          "properties": {
+            "action": {
+              "description": "Describes which action should be triggered by the bot category.",
+              "$ref": "#/definitions/Action"
+            }
+          }
+        },
+        "bad": {
+          "description": "Describes the action to be taken when detecting a bad bot.",
+          "properties": {
+            "action": {
+              "description": "Describes which action should be triggered by the bot category.",
+              "$ref": "#/definitions/Action"
+            }
+          }
+        },
+        "overrides": {
+          "description": "List of custom bot categories which take precedence over the predefined categories.",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/CustomBotCategory"
+          }
+        }
+      }
+    },
+    "BotIdentifier": {
+      "description": "Defines a way to capture a bot request.",
+      "required": [
+        "type",
+        "value",
+        "matchOperator"
+      ],
+      "properties": {
+        "type": {
+          "description": "Describes which property of the request to match against.",
+          "type": "string",
+          "enum": [
```

Comment:
Should we also be able to match on other fields? What is the difference/advantage between making a custom bot category and making a custom rule?

Reply:
I thought that was one of the requirements Eric and Teresa had come up with. I just IM'd Eric to make sure.

The custom bot category acts as an override for the existing categories. If a custom bot category gets created with the "Allow" action, then that would take precedence over, say, the Bad bot category. Unlike the custom rule, it wouldn't make the entire request skip the rest of the EvalRequest pipeline.

Reply:
Just checked with Eric, seems like this should mimic the CustomRule behavior, therefore I'll opt for re-using `MatchCondition`